contrib/errorprone/src/python/pants/contrib/errorprone/tasks/errorprone.py

```diff
@@ -0,0 +1,129 @@
+                  'specified on the command line.')
+    register('--excluded-targets', type=list, default=[], fingerprint=True,
+             help='Targets to be excluded from analysis.')
+    register('--command-line-options', type=list, default=[], fingerprint=True,
+             help='Command line options passed to Error Prone')
+
+    cls.register_jvm_tool(register,
+                          'errorprone',
+                          classpath=[
+                            JarDependency(org='com.google.errorprone',
+                                          name='error_prone_core',
+                                          rev='2.0.15'),
+                          ],
+                          main=cls._ERRORPRONE_MAIN,
+                          custom_rules=[
+                            Shader.exclude_package('com.google.errorprone', recursive=True),
+                          ]
+                         )
+
+  @classmethod
+  def prepare(cls, options, round_manager):
+    super(ErrorProne, cls).prepare(options, round_manager)
+    round_manager.require_data('runtime_classpath')
+
+  def _is_errorprone_target(self, target):
+    return target.has_sources(self._JAVA_SOURCE_EXTENSION) and not target.is_synthetic
+
+  @property
+  def cache_target_dirs(self):
+    return True
+
+  def execute(self):
+    if self.get_options().skip:
+      return
+
+    if self.get_options().transitive:
+      targets = self.context.targets(self._is_errorprone_target)
+    else:
+      targets = filter(self._is_errorprone_target, self.context.target_roots)
+
+    targets = list(set(targets))
+
+    target_count = 0
+    errorprone_failed = False
+    with self.invalidated(targets, invalidate_dependents=True) as invalidation_check:
+      total_targets = len(invalidation_check.invalid_vts)
+      for vt in invalidation_check.invalid_vts:
+        target_count += 1
+        self.context.log.info('[{}/{}] {}'.format(str(target_count).rjust(len(str(total_targets))), total_targets, vt.target.address.spec))
+
+        if vt.target.address.spec in self.get_options().excluded_targets:
+          self.context.log.info('  Skipping [{}] because it is an excluded target'.format(vt.target.address.spec))
+          vt.update()
+          continue
+
+        result = self.errorprone(vt.target, self.calculate_sources(vt.target))
+        if result != 0:
+          if self.get_options().fail_fast:
+            raise TaskError('java {main} ... exited non-zero ({result})'.format(
+                main=self._ERRORPRONE_MAIN, result=result))
+          else:
+            errorprone_failed = True
+        else:
+          vt.update()
+
+      if errorprone_failed:
+        raise TaskError('ErrorProne checks failed')
+
+  def calculate_sources(self, target):
+    return {source for source in target.sources_relative_to_buildroot()
+            if source.endswith(self._JAVA_SOURCE_EXTENSION)}
+
+  def errorprone(self, target, sources):
```

Comment:
Good call.  I'll move the `self.calculate_sources(target)` into the findbugs method.